### ISSUE TITLE
[DataGridPro] Keyboard navigation doesn't work as expected in header filters

### ISSUE THREAD

**Issue by MBilalShafi** [mui/mui-x#10357]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Steps to reproduce 🕹

Link to live example: https://mui.com/x/react-data-grid/filtering/header-filters/

Steps:
1. Click/focus on any data cell
2. Navigate to a header filter cell using keyboard arrow keys
3. Press Enter a few times.

### Current behavior 😯

After Toggling Editing state on, the editing state is stuck to edit mode and further pressing Enter key has no visual update

![focus-issue](https://github.com/mui/mui-x/assets/12609561/7fccd171-cd6e-4fb7-b53a-8701a10209a4)

### Expected behavior 🤔

Pressing Enter key again and again should toggle the editing state for a cell to on and off

### Context 🔦

Discovered while investigating https://github.com/mui/mui-x/issues/10325

This issue was arisen due to this change: https://github.com/mui/mui-x/pull/9181/files#diff-13450384ad8e3493a92f90696c4b890a204aae9e0099ec7647cd223ed9d22408R450 which fixed one issue (7.1 in https://github.com/mui/mui-x/issues/9034#issuecomment-1570343195) but created another (mentioned in this ticket)

### Your environment 🌎

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>
